ports/esp32/esp32_pcnt.c

```diff
@@ -0,0 +1,1231 @@
+ @param neg_mode Counter mode when detecting negative edge
+ @param hctrl_mode Counter mode when control signal is high level
+ @param lctrl_mode Counter mode when control signal is low level
+
+ @note
+     Can raise EspException:
+     - ESP_ERR_INVALID_STATE pcnt driver has not been initialized
+     - ESP_ERR_INVALID_ARG Parameter error
+*/
+/*
+STATIC mp_obj_t pcnt_PCNT_set_mode(size_t n_args, const mp_obj_t *args) {
+    pcnt_PCNT_obj_t *self = MP_OBJ_TO_PTR(args[0]);
+    mp_int_t channel = mp_obj_get_int(args[1]);
+    mp_int_t pos_mode = mp_obj_get_int(args[2]);
+    mp_int_t neg_mode = mp_obj_get_int(args[3]);
+    mp_int_t hctrl_mode = mp_obj_get_int(args[4]);
+    mp_int_t lctrl_mode = mp_obj_get_int(args[5]);
+
+    check_esp_err(pcnt_set_mode(self->unit, channel, pos_mode, neg_mode, hctrl_mode, lctrl_mode));
+
+    return MP_ROM_NONE;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(pcnt_PCNT_set_mode_obj, 6, 6, pcnt_PCNT_set_mode);
+*/
+
+// def PCNT.set_pin(self, channel: int, pulse_io: int, ctrl_io: int)
+/*
+Configure PCNT pulse signal input pin and control input pin
+
+ @param channel PCNT channel number
+ @param pulse_io Pulse signal input GPIO
+ @param ctrl_io Control signal input GPIO
+
+Set the signal input to PCNT_PIN_NOT_USED if unused.
+
+ @note
+     Can raise EspException:
+     - ESP_ERR_INVALID_STATE pcnt driver has not been initialized
+     - ESP_ERR_INVALID_ARG Parameter error
+*/
+/*
+STATIC mp_obj_t pcnt_PCNT_set_pin(size_t n_args, const mp_obj_t *args) {
+    pcnt_PCNT_obj_t *self = MP_OBJ_TO_PTR(args[0]);
+    mp_int_t channel = mp_obj_get_int(args[1]);
+    mp_int_t pulse_io = mp_obj_get_int(args[2]);
+    mp_int_t ctrl_io = mp_obj_get_int(args[3]);
+
+    check_esp_err(pcnt_set_pin(self->unit, channel, pulse_io, ctrl_io));
+
+    return MP_ROM_NONE;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(pcnt_PCNT_set_pin_obj, 4, 4, pcnt_PCNT_set_pin);
+*/
+
+// def PCNT.unit_config(self, pcnt_config: int)
+/*
+Configure Pulse Counter unit
+ @note
+  This function will disable three events: PCNT_EVT_L_LIM, PCNT_EVT_H_LIM, PCNT_EVT_ZERO.
+
+ @param pcnt_config Pointer of Pulse Counter unit configure parameter
+
+ @note
+     Can raise EspException:
+     - ESP_ERR_INVALID_STATE pcnt driver already initialized
+     - ESP_ERR_INVALID_ARG Parameter error
+*/
+/*
+STATIC mp_obj_t pcnt_PCNT_unit_config(mp_obj_t self_obj, mp_obj_t pcnt_config_obj) {
+    // pcnt_PCNT_obj_t *self = MP_OBJ_TO_PTR(self_obj);
+    pcnt_config_t *pcnt_config = MP_OBJ_TO_PTR(pcnt_config_obj);
+
+    check_esp_err(pcnt_unit_config(pcnt_config));
+
+    return MP_ROM_NONE;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_2(pcnt_PCNT_unit_config_obj, pcnt_PCNT_unit_config);
+*/
+// ====================================================================================
+STATIC mp_obj_t pcnt_PCNT_set_count(mp_obj_t self_obj, mp_obj_t value_obj) {
+    pcnt_PCNT_obj_t *self = MP_OBJ_TO_PTR(self_obj);
+    int64_t value = mp_obj_get_int(value_obj);
```

Comment:
mp_obj_get_int() returns a 32bit value on ESP32. So you cannot set the count to the full range.

Reply:
@robert-hh Thanks for your attentiveness, you are right. 
Is there an example of passing int64 somewhere?

Reply:
Since I had the same demand, I searched in the MP code for a while and did not find something suitable. So I sketched a little function myself. It is pretty simple and does not much of checking, especially not for overflow. So take it with care (or add a len check).
```

STATIC uint64_t mp_get_ll_int(mp_obj_t arg_in) {
    uint64_t res;
    if (mp_obj_is_small_int(arg_in)) {
        res = MP_OBJ_SMALL_INT_VALUE(arg_in);
    } else if (mp_obj_is_type(arg_in, &mp_type_int)) {
        mp_obj_int_t *arg = arg_in;
        int len = arg->mpz.len;
        res = 0;
        for (int i = len - 1; i >= 0; i--) {
            res = (res << MPZ_DIG_SIZE) + arg->mpz.dig[i];
        }
        if (arg->mpz.neg) {
            res = -res;
        }
    } else {
        mp_raise_ValueError(MP_ERROR_TEXT("Argument must be a number")); 
    }
    return res;
}
```